### ISSUE TITLE
Instructions for manual installation of xhyve driver do not work on Sierra

### ISSUE THREAD

**Issue by agajdosi** [minishift/minishift#1123]
I have followed documented instructions for [manual installation of xhyve driver](https://docs.openshift.org/latest/minishift/getting-started/setting-up-driver-plugin.html#manual-install) and was then unable to start minishift properly:

```
cdkqe-mac-01:bin hudson$ sudo curl -L  https://github.com/zchee/docker-machine-driver-xhyve/releases/download/v0.3.1/docker-machine-driver-xhyve -o /usr/local/bin/docker-machine-driver-xhyve
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   617    0   617    0     0    453      0 --:--:--  0:00:01 --:--:--   453
100 12.6M  100 12.6M    0     0  1126k      0  0:00:11  0:00:11 --:--:-- 2118k
cdkqe-mac-01:bin hudson$ sudo chown root:wheel /usr/local/bin/docker-machine-driver-xhyve

cdkqe-mac-01:bin hudson$ ls -l docker-machine-driver-xhyve 
-rw-r--r--  1 root  wheel  13241508 12 Jul 00:30 docker-machine-driver-xhyve
cdkqe-mac-01:bin hudson$ sudo chmod u+s /usr/local/bin/docker-machine-driver-xhyve
cdkqe-mac-01:bin hudson$ ls -l docker-machine-driver-xhyve 
-rwSr--r--  1 root  wheel  13241508 12 Jul 00:30 docker-machine-driver-xhyve
cdkqe-mac-01:bin hudson$ which docker-machine-driver-xhyve 

cdkqe-mac-01:bin hudson$ minishift start --vm-driver xhyve
Starting local OpenShift cluster using 'xhyve' hypervisor...
E0712 00:42:27.786796   15896 start.go:275] Error starting the VM: Driver "xhyve" not found. Do you have the plugin binary "docker-machine-driver-xhyve" accessible in your PATH?. Retrying.
E0712 00:42:27.787557   15896 start.go:275] Error starting the VM: Driver "xhyve" not found. Do you have the plugin binary "docker-machine-driver-xhyve" accessible in your PATH?. Retrying.
E0712 00:42:27.787588   15896 start.go:275] Error starting the VM: Driver "xhyve" not found. Do you have the plugin binary "docker-machine-driver-xhyve" accessible in your PATH?. Retrying.
Error starting the VM: Driver "xhyve" not found. Do you have the plugin binary "docker-machine-driver-xhyve" accessible in your PATH?
Driver "xhyve" not found. Do you have the plugin binary "docker-machine-driver-xhyve" accessible in your PATH?
Driver "xhyve" not found. Do you have the plugin binary "docker-machine-driver-xhyve" accessible in your PATH?
```

Since using brew for installation Minishift works as expected.

When discussing on Minishift IRC channel @gbraad also pointed out that there are not any instructions for installation of actual xhyve - could be `brew install --HEAD xhyve` which mentions xhyve documentation. Whole section is focused on setting up the driver plugins but as parts for Fedora and Debian/Ubuntu mention actual installation of kvm we could think of aligning xhyve part with it also.

**Comment by praveenkumar** [comment]
Problem is our manual instructions doesn't put details about how to install dependencies for `docker-machine-driver-xhyve` and that's reason it fails. I don't mind to remove this manual instruction completely and only rely on `brew install docker-machine-driver-xhyve` because then it will pull down all required dependencies.

**Comment by hferentschik** [comment]
@agajdosi, so where is the binary when you install it via brew? And is _/usr/local/bin_ really on our PATH? Looking at the output of your `which docker-machine-driver-xhyve` it does not seem so.

**Comment by agajdosi** [comment]
@hferentschik It is in same location:
```
cdkqe-mac-01:~ hudson$ which docker-machine-driver-xhyve
/usr/local/bin/docker-machine-driver-xhyve
```

**Comment by hferentschik** [comment]
> Problem is our manual instructions doesn't put details about how to install dependencies for docker-machine-driver-xhyve

Which are? I have no dependencies installed for what I can tell. Definitely not the xhyve binary. For all I know the docker-machine-driver-xhyve interacts with Hypervisor.framework (which is native to macOS) directly. I have not _xhyve_ binary on my machine.

>  I don't mind to remove this manual instruction completely and only rely on brew install docker-machine-driver-xhyve because then it will pull down all required dependencies.

well, not all macOS users use brew. I am a Mac Port user. I prefer to install manually. For all I can tell, there is nothing wrong with the manual steps.

**Comment by hferentschik** [comment]
> It is in same location

So why in your original output did `which docker-machine-driver-xhyve` not show anything. Could it be that you did originally not have _/usr/local/bin_ on your PATH, but the brew install added it, and hence it worked?

**Comment by LalatenduMohanty** [comment]
I do not have Xhyve installed on my macOS. @agajdosi After installing Xhyve it started working for you?

**Comment by hferentschik** [comment]
Looking at the commands you ran and their output, as well as the actual error message all points to that Minishift could not find the driver binary.

**Comment by hferentschik** [comment]
@LalatenduMohanty Pretty sure it is the PATH. Assuming @agajdosi has neither use brew or Mac Ports, he might not have had '/usr/local/bin' on the PATH. I this case you would see the above behaviour, even so the driver is there. 

When using brew it also sets up the PATH is required (same as Mac Ports).

**Comment by agajdosi** [comment]
> So why in your original output did which docker-machine-driver-xhyve not show anything. Could it be that you did originally not have /usr/local/bin on your PATH, but the brew install added it, and hence it worked?

@hferentschik  I have checked with `echo $PATH` for that case and /usr/local/bin was on path. Also git binary was lying in same folder next to the `docker-machine-driver-xhyve` and was detected by `which`.

```
cdkqe-mac-01:bin hudson$ echo $PATH
/Users/hudson/minishift:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/go/bin
```

> Pretty sure it is the PATH.

> When using brew it also sets up the PATH is required (same as Mac Ports).

Unfortunately I have installed xhyve binary via brew and then started to take care of xhyve driver so even if I am not counting echo check, it had to be on PATH.

> Which are? I have no dependencies installed for what I can tell. 

Brew installs docker-machine, ocaml, ocamlbuild, camlp4, gringo, clasp, aspcud, opam, libev: 
```
cdkqe-mac-01:bin hudson$ brew install docker-machine-driver-xhyve
==> Installing dependencies for docker-machine-driver-xhyve: docker-machine, ocaml, ocamlbuild, camlp4, gringo, clasp, aspcud, opam, libev
```

@LalatenduMohanty Can verify that xhyve binary is not needed.

Going to uninstall xhyve driver and repeat the manual install to see the result.

**Comment by hferentschik** [comment]
> Going to uninstall xhyve driver and repeat the manual install to see the result.

+1

**Comment by gbraad** [comment]
> Which are? I have no dependencies installed for what I can tell

The dependeny you get with `brew install` is `xhyve` itself. In the document we do not mention how to install this.

**Comment by agajdosi** [comment]
```
cdkqe-mac-01:~ hudson$ brew uninstall docker-machine-driver-xhyve
Uninstalling /usr/local/Cellar/docker-machine-driver-xhyve/0.3.3... (3 files, 10.3MB)
cdkqe-mac-01:~ hudson$ which docker-machine-driver-xhyve
cdkqe-mac-01:~ hudson$ sudo curl -L  https://github.com/zchee/docker-machine-driver-xhyve/releases/download/v0.3.1/docker-machine-driver-xhyve -o /usr/local/bin/docker-machine-driver-xhyve
Password:
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   617    0   617    0     0    300      0 --:--:--  0:00:02 --:--:--   300
100 12.6M  100 12.6M    0     0   489k      0  0:00:26  0:00:26 --:--:--  367k
cdkqe-mac-01:~ hudson$ sudo chown root:wheel /usr/local/bin/docker-machine-driver-xhyve
cdkqe-mac-01:~ hudson$ sudo chmod u+s /usr/local/bin/docker-machine-driver-xhyve

cdkqe-mac-01:~ hudson$ minishift start
-- Installing default add-ons ... OK
Starting local OpenShift cluster using 'xhyve' hypervisor...
Downloading ISO 'https://github.com/minishift/minishift-b2d-iso/releases/download/v1.0.2/minishift-b2d.iso'
 40.00 MiB / 40.00 MiB [================================================================] 100.00% 0s
E0712 02:25:38.967047   18474 start.go:275] Error starting the VM: Driver "xhyve" not found. Do you have the plugin binary "docker-machine-driver-xhyve" accessible in your PATH?. Retrying.
E0712 02:25:38.967412   18474 start.go:275] Error starting the VM: Driver "xhyve" not found. Do you have the plugin binary "docker-machine-driver-xhyve" accessible in your PATH?. Retrying.
E0712 02:25:38.967446   18474 start.go:275] Error starting the VM: Driver "xhyve" not found. Do you have the plugin binary "docker-machine-driver-xhyve" accessible in your PATH?. Retrying.
Error starting the VM: Driver "xhyve" not found. Do you have the plugin binary "docker-machine-driver-xhyve" accessible in your PATH?
Driver "xhyve" not found. Do you have the plugin binary "docker-machine-driver-xhyve" accessible in your PATH?

cdkqe-mac-01:~ hudson$ which docker-machine-driver-xhyve
cdkqe-mac-01:~ hudson$ echo $PATH
/Users/hudson/minishift:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/go/bin
```

Feels unfair...

**Comment by hferentschik** [comment]
> The dependency you get with brew install is xhyve itself. In the document we do not mention how to install this.

I don't have the xhyve binary, neither does @LalatenduMohanty. I believe docker-machine-driver-xhyve interacts with Hypervisor.framework directly.

**Comment by agajdosi** [comment]
I have also uninstalled xhyve binary and it works for me. Brew does not mention xhyve as dependency when installing docker-machine-driver-xhyve.

**Comment by gbraad** [comment]
I have it definitely installed.

According to https://github.com/zchee/docker-machine-driver-xhyve/tree/master/vendor/github.com/zchee/libhyperkit (vendor library), it does not interact directly with hypervisor.framework, but exposes a binding to xhyve. maybe the brew module compiles it corectly: https://github.com/Homebrew/homebrew-core/blob/master/Formula/docker-machine-driver-xhyve.rb while manual needs to interact with the binary ?

**Comment by LalatenduMohanty** [comment]
Wondering if having Docker for mac (which I have) has anything to do with this.

**Comment by agajdosi** [comment]
So I guess execution rights are needed. Tried to add `sudo chmod +x /usr/local/bin/docker-machine-driver-xhyve` after the download and now I am able to start minishift.

Now trying to figure out if dependencies installed by brew are needed.

**Comment by LalatenduMohanty** [comment]
@gbraad You should uninstall it and it should still work for you.

**Comment by gbraad** [comment]
also consider to look into Docker for Mac, as I also have this installed.

**Comment by gbraad** [comment]
@LalatenduMohanty I actually use xhyve standalone with xhyve-manager to run other VMs. Uninstall is not an option...

**Comment by gbraad** [comment]
> https://github.com/zchee/docker-machine-driver-xhyve/tree/master/vendor/github.com/zchee/libhyperkit

This is actually a rename and fork from Michael Steil's `xhyve`: you can see this from the `README.xhyve.md`

**Comment by agajdosi** [comment]
So nothing from `docker-machine, ocaml, ocamlbuild, camlp4, gringo, clasp, aspcud, opam, libev` is needed for manual installation of docker-machine-driver-xhyve.

The only problem were missing execution rights on docker-machine-driver-xhyve binary. I have tested following steps for manual install:
1. curl -L  https://github.com/zchee/docker-machine-driver-xhyve/releases/download/v0.3.1/docker-machine-driver-xhyve -o /usr/local/bin/docker-machine-driver-xhyve
2. chmod +x /usr/local/bin/docker-machine-driver-xhyve
3. chown root:wheel /usr/local/bin/docker-machine-driver-xhyve
4. sudo chmod u+s /usr/local/bin/docker-machine-driver-xhyve

From my findigs on Sierra there is no need for curl with sudo, this is also mentioned on release of [docker-machine-driver-xhyve v0.3.1](https://github.com/zchee/docker-machine-driver-xhyve/releases/tag/v0.3.1). Next step is adding execution rights. Last two steps are identical.

Based on @gbraad findings it could be maybe good to add a sentence about fact, that machine-docker-driver-xhyve vendors xhyve in itself and thus there is no need for installation of xhyve binary.

**Comment by hferentschik** [comment]
> Next step is adding execution rights. Last two steps are identical with current docs.

So making the binary executable is the missing bit then, right? 
I think it is less about adding a step, but rather fixing step 4:

    sudo chmod u+xs /usr/local/bin/docker-machine-driver-xhyve

The binary must be executable _and_ run as root.

> Based on @gbraad findings it could be maybe good to add a sentence about fact, that machine-docker-driver-xhyve contains/vendors xhyve in itself and thus there is no need for installation of xhyve binary.

Not so sure about this. Why making things more complicated? This is an implementation detail a user should not have to know/think about.

**Comment by gbraad** [comment]
> Not so sure about this. Why making things more complicated? This is an
implementation detail a user should not have to know/think about.

I see people also installing docker-machine on macOS and Linux. Being a
little bit more informative is not bad.

-- 

   Gerard Braad | http://gbraad.nl
   [ Doing Open Source Matters ]

**Comment by hferentschik** [comment]
> I see people also installing docker-machine on macOS and Linux. Being a little bit more informative is not bad.

There is a line somewhere, but sure.

**Comment by thatdocslady** [comment]
Hi all, going through the doc issues to see where I can help. If I understand this whole thread correctly, basically what needs to be fixed is the content in this section:

https://github.com/minishift/minishift/blob/master/docs/source/getting-started/setting-up-driver-plugin.adoc#manual-install

@hferentschik from what I can see the command that you want to change is in step 3, not step 4. This is the existing text:

> Set owner User ID (SUID) for the binary as follows:
> $ sudo chmod u+s /usr/local/bin/docker-machine-driver-xhyve

And from what I understand, you want to change the command to this syntax:
> $ sudo chmod u+xs /usr/local/bin/docker-machine-driver-xhyve

@gbraad, it looks like step 2 of this section already covers the root access to the driver, but please clarify if you think we should provide additional info about it.

Let me know if you want me to pick up this fix based on what I wrote, or if you have any plans to fix this in another way.

**Comment by agajdosi** [comment]
Tested it today @hferentschik and it seems that to set `-rwsr-xr-x` for docker-machine-driver-xhyve, which I believe is what we need, one has to run: `sudo chmod u+s,+x /usr/local/bin/docker-machine-driver-xhyve`. This I have tested and found to work fine. With `u+xs` there are missing execution rights for others, which is same as for original `u+s`.

So from my side, @thatdocslady, please change the command for: `sudo chmod u+s,+x /usr/local/bin/docker-machine-driver-xhyve`.

Last question is if we want to also include short information about docker-machine being packed in docker-machine-driver-xhyve, so there is no need to have solo docker-machine binary at all. @gbraad?

**Comment by gbraad** [comment]
No additional information needed.
as @hferentschik mentioned, he rather leaves the docker-machine reference out.

If people install this, they have done so on their own... which means they haven't followed documentation

**Comment by hferentschik** [comment]
Let's just change the instructions.

**Comment by hferentschik** [comment]
Merged via pull request #1155
